src/Introspection.hack

```diff
@@ -0,0 +1,289 @@
+    public function getFields(bool $include_deprecated = false): ?vec<IntrospectableField>;
+}
+
+interface IntrospectableObject extends IntrospectableTypeWithFields {
+    public function getFields(bool $include_deprecated = false): ?vec<IntrospectableField>;
+    public function getInterfaces(): ?vec<IntrospectableInterface>;
+    public static function nonNullable(): IntrospectableType;
+    public static function nullable(): IntrospectableType;
+    public static function literal(): IntrospectableType;
+}
+
+interface IntrospectableInterface extends IntrospectableTypeWithFields {
+    public function getFields(bool $include_deprecated = false): ?vec<IntrospectableField>;
+    public function getPossibleTypes(): ?vec<IntrospectableType>;
+}
+
+interface IntrospectableEnum extends IntrospectableType {
+    public function getEnumValues(bool $include_deprecated = false): ?vec<IntrospectableEnumValue>;
+}
+
+interface IntrospectableInput extends IntrospectableType {
+    public function getInputFields(bool $include_deprecated = false): ?vec<IntrospectableInputValue>;
+}
+
+interface IntrospectableGeneric {
+    public function getInnerType(): IntrospectableType;
+}
+
+interface IntrospectableNamedType {
+    public function getName(): string;
+    public function getDescription(): ?string;
+}
+
+interface IntrospectableField extends IntrospectableNamedType {
+    public function getArgs(): vec<IntrospectableInputValue>;
+    public function getType(): IntrospectableType;
+    public function isDeprecated(): bool;
+    public function getDeprecationReason(): ?string;
+}
+
+interface IntrospectableInputValue extends IntrospectableNamedType {
+    public function getType(): IntrospectableType;
+    public function defaultValue(): ?string;
+}
+
+interface IntrospectableEnumValue extends IntrospectableNamedType {
+    public function isDeprecated(): bool;
+    public function getDeprecationReason(): ?string;
+}
+
+final abstract class IntrospectionQueryFields {
+    // filter out the schema argument if it is an introspection field?
+    <<\Slack\GraphQL\QueryRootField('__schema', 'Introspect the schema')>>
+    public static function getSchema(classname<IntrospectableSchema> $schema): ?__Schema {
+        return new __Schema($schema);
+    }
+
+    <<\Slack\GraphQL\QueryRootField('__type', 'Introspect a specific type')>>
+    public static function getType(classname<IntrospectableSchema> $schema, string $name): ?__Type {
+        $type = $schema::getType($name);
+        return $type is nonnull ? new __Type($type) : null;
+    }
+}
+
+<<\Slack\GraphQL\EnumType('__TypeKind', '__TypeKind')>>
+enum __TypeKind: string {
+    SCALAR = 'SCALAR';
+    OBJECT = 'OBJECT';
+    INTERFACE = 'INTERFACE';
+    UNION = 'UNION';
+    ENUM = 'ENUM';
+    INPUT_OBJECT = 'INPUT_OBJECT';
+    LIST = 'LIST';
+    NON_NULL = 'NON_NULL';
+}
+
+<<\Slack\GraphQL\ObjectType('__Schema', 'Schema introspection')>>
```

Comment:
I used these wrapping types to prevent these fields from being queryable in the objects themselves. I'm not sure if there is a simpler way to do this. I'm going to look with some fresh eyes tomorrow 😄

Reply:
> to prevent these fields from being queryable in the objects themselves

Hmm I don't quite understand this, what do you mean "queryable in the objects themselves"?

Having all these wrappers that mostly just delegate to the inner class really does seem like a lot of boilerplate code that should be avoidable somehow...

Reply:
I was initially thinking we could do something like:

```
<<\Slack\GraphQL\InterfaceType('__Type', 'Introspectable type')>>
interface __Type {
...
}
```

with:

```
<<\Slack\GraphQL\ObjectType('User', 'User')>>
final class User implements __Type {
...
}
```

but then you'd be able to query:

```
{
    user {
       // introspection fields
       name
       kind
       fields
       ...etc
    }
}
```

which wouldn't make sense?

Reply:
OK so I think this is a bit confusing because the classes have the same name (only different namespace) but I think we're talking about 2 different classes:

- `\User` is the class that has `<<GraphQL\Object('User')>>` on it and has all the non-introspection field methods (`<<GraphQL\Field('favorite_color')>>` etc.)
- `\Slack\GraphQL\Test\Generated\User` is the class we generate from `\User`; this is the class that has all the introspection methods like `<<GraphQL\Field('kind')>>` (probably inherited from IntrospectableType, not actually part of the generated code itself)